MRG: New time gen class

dengemann opened the pull request:
Closes #1302. 

@kingjr sorry rebasing was a mess, I had to squash your commits to get started within a finite amount of time. Your commits are supposed to be logged and counted anyways

Here's a list of things I addressed.

1. Enforce sklaern API that distinguishes between parametersa and attributes
2. make score and predict return the corresponding values
3. fix unfarable expression which create references not copies e.g ```[[]] * len(test_times``` creates a list of len(test_times) times the same list object.
4. fix the default y and simplify the example. I figured out that sklearn apparently expects 0 and 1 or True and False for the binary case. Using scipy.stats.rankdata as a trick to achieve this.
5. isolated the changes related to the new class from the changes related to the previous attempt to modify the existing function.
6. rebased and fixed unit tests.
7. caught a corner case where predict == 'distance' will crash (binary casses).
8. addressed pep8 issues
9. added doc to better document for developers and users the attributes. e.g. the somewhat complex list of lists approach to scores and estimators (yes I know not easy to do otherwise).

Todo:

1. doc is not fully complete.
2. Check a few parts in the code with you @kingjr to check whetehr we can do bettr.
3. extend tests.
4. update whats new once we're don here.

Important:

Absolute feature add block. We're going to stabilize what's there and then we move on in *smaller* steps. For this @kingjr I only want to hear from you that the overall structure is expected to stay.

@agramfort if you're looking for a review job ...

Eric89GXL commented:
What's the status here?

choldgraf commented:
If there is extra testing etc, I can still help with this.

agramfort commented:
any help in testing, code quality reviews, and reports of timings on actual data is welcome

choldgraf commented:
Hey all - I've been playing around with the time gen class. It seems like there are some configurations under which it is throwing errors. I'm not really sure why as of yet, but I keep getting the following IndexError when trying to predict:

```python
/home/knight/holdgraf/github/public/mne_time_gen/mne/decoding/time_gen.pyc in _predict_time_loop(X, estimators, cv, slices, predict_mode, predict_type)
    473                     y_pred[t] = np.empty((n_trial, y_pred_.shape[1]))
    474                     y_pred[t][test, :] = y_pred_
--> 475                 y_pred[t][test, :] = _predict(Xtrain[test, :],
    476                                               estimators[k:k + 1],
    477                                               predict_type)

IndexError: index 12 is out of bounds for axis 0 with size 12
```

Have you guys run into any issues like these? I'm basically using the defaults for everything here, on a signal that is sampled at 1000Hz with -.1 to 2 seconds for each epoch. In this case, I'm training on one subset of trials and trying to predict on another. I can try to pry into this more later to figure out what may be causing the error, but wanted to see if this was something known already.

Eric89GXL commented:
`pdb` is your friend, it has lots of useful commands:
```
>>> import pdb; pdb.pm()  # start post-mortem debugging
(pdb) u  # to move up in the stack trace (if you want)
(pdb) d  # to move down in the stack trace (if you want)
(pdb) p Xtrain.shape  # print a variable
(pdb) p test
```
If you're using a nice IDE like Spyder, as `pdb` moves up and down the stack trace it will open each offending file and move to the proper line.

choldgraf commented:
Yup - I'm planning to look into it tomorrow, but I had to leave work and
felt like throwing this out there before I left :P hopefully I can post
something more useful tomorrow haha

On Thu, Feb 12, 2015 at 7:14 PM, Eric Larson <notifications@github.com>
wrote:

> pdb is your friend, it has lots of useful commands:
>
> >>> import pdb; pdb.pm()  # start post-mortem debugging
> (pdb) u  # to move up in the stack trace (if you want)
> (pdb) d  # to move down in the stack trace (if you want)
> (pdb) p Xtrain.shape  # print a variable
> (pdb) p test
>
> If you're using a nice IDE like Spyder, as pdb moves up and down the
> stack trace it will open each offending file and move to the proper line.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1629#issuecomment-74198519>.
>

choldgraf commented:
Ah I think I figured it out - when you fit the time gen object, it creates a cv object and assigns it to self.cv_

In my use case, I was trying to use a subset of trials to train an test. The problem with this is that the `score` method loops through `self.cv_` if you don't supply one, and since I was trying to score on a different number of trials than I tested on, it was indexing incorrectly. Basically, it seems like the fit/score methods are designed to work in tandem on the same trial dataset. It is possible to pass other objects to it, but then you'd also need to pass a custom cv object as well.

kingjr commented:
Yes the fit and score functions do work in tandem. If you want to pass
different trials to predict or score function and avoid creating a new cv
object, you can use

 gat.predict_mode = 'mean-prediction'

This will generate a single prediction per new trial by averaging the
prediction of each cv classifier.

Would this address your use case?

JR

On Saturday, 14 February 2015, Chris Holdgraf <notifications@github.com>
wrote:

> Ah I think I figured it out - when you fit the time gen object, it creates
> a cv object and assigns it to self.cv_
>
> In my use case, I was trying to use a subset of trials to train an test.
> The problem with this is that the score method loops through self.cv_ if
> you don't supply one, and since I was trying to score on a different number
> of trials than I tested on, it was indexing incorrectly. Basically, it
> seems like the fit/score methods are designed to work in tandem on the same
> trial dataset. It is possible to pass other objects to it, but then you'd
> also need to pass a custom cv object as well.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1629#issuecomment-74344947>.
>

choldgraf commented:
It sounds like this would work, but there should be some kind of warning
either in the docstring or an assertion or something, since it was unclear
why this wasn't working before.

On Fri, Feb 13, 2015 at 5:05 PM, J-R King <notifications@github.com> wrote:

> Yes the fit and score functions do work in tandem. If you want to pass
> different trials to predict or score function and avoid creating a new cv
> object, you can use
>
> gat.predict_mode = 'mean-prediction'
>
> This will generate a single prediction per new trial by averaging the
> prediction of each cv classifier.
>
> Would this address your use case?
>
> JR
>
> On Saturday, 14 February 2015, Chris Holdgraf <notifications@github.com>
> wrote:
>
> > Ah I think I figured it out - when you fit the time gen object, it
> creates
> > a cv object and assigns it to self.cv_
> >
> > In my use case, I was trying to use a subset of trials to train an test.
> > The problem with this is that the score method loops through self.cv_ if
> > you don't supply one, and since I was trying to score on a different
> number
> > of trials than I tested on, it was indexing incorrectly. Basically, it
> > seems like the fit/score methods are designed to work in tandem on the
> same
> > trial dataset. It is possible to pass other objects to it, but then you'd
> > also need to pass a custom cv object as well.
> >
> > --
> > Reply to this email directly or view it on GitHub
> > <https://github.com/mne-tools/mne-python/pull/1629#issuecomment-74344947
> >.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1629#issuecomment-74353511>.
>

dengemann commented:
@agramfort @choldgraf @Eric89GXL with @kingjr we're done, we think, hope, expect.

choldgraf commented:
great!

dengemann commented:
some pep8 issues are left ... apart from that I'd really like to merge this PR. Each month it's up in the air generates more and more work.

Eric89GXL commented:
A couple of @kingjr's comments still remain (some wording changes). I'll try to take a more substantive look tomorrow.

agramfort commented:
let me know when everybody is happy so I can be the bad guy :)

dengemann commented:
@agramfort no more bad guys please :) Some flakes are left to be addressed. @kingjr amd I am happy with it. The comments and XXX left in the code are for future PRs, they go to far for now. This code will be used by many people, so bug fixes are expected as the code is actually used.

choldgraf commented:
I am excited to be able to use this code from the main MNE repository instead of loading Denis' branch every time :) 

There are some extensions that I'd like to propose at a later time (e.g., have you guys thought about "frequency" generalization rather than just time generalization?), but personally this is already a great improvement.

dengemann commented:
Travis is happy

dengemann commented:
I need to check a potential issue with tests on the current sklearn master

Eric89GXL commented on mne/decoding/time_gen.py:
```diff
@@ -1,14 +1,984 @@
+            'predict' : generates a categorical estimate of each trial.
+            'predict_proba' : generates a probabilistic estimate of each trial.
+            'decision_function' : generates a continuous non-probabilistic
+                estimate of each trial.
```
Do these look okay in the doc builds? They might need an extra newline before the new section and/or bullets

dengemann replied:
good call. let me run a doc build

Eric89GXL commented on mne/decoding/time_gen.py:
```diff
@@ -1,14 +1,984 @@
+
+    Attributes
+    ----------
+    y_train_ : np.ndarray, shape (n_samples,)
```
Why put an underscore after all these names? Are we adopting a new convention?

dengemann replied:
scikit learn convention for attributes created by .fit

dengemann replied:
as compared to constructor

Eric89GXL commented on mne/decoding/time_gen.py:
```diff
@@ -1,14 +1,984 @@
+            Defaults to None
+        picks : np.ndarray of int, shape (n_channels,) | None
+            The channels to be used. If None, defaults to meg and eeg channels.
+            Defaults to None.
```
undocumented return

Eric89GXL commented on mne/utils.py:
```diff
@@ -1685,3 +1685,33 @@ def _sphere_to_cartesian(theta, phi, r):
+
+def create_slices(start, stop, step=None, length=1):
+    """ Generate slices of time indexes
+    Parameters
```
missing newline

Eric89GXL commented:
Other than my comments, this looks reasonable to me. The example ran fine on my machine. There is a lot of code to get this functionality, though, and I don't totally understand it. I'm +1 for merge once @agramfort takes a close look.

dengemann commented:
@Eric89GXL addressed your comments. For the doc I cannot really proceed, fighting with sphinx which seems terribly broken for some reasons on my machine.

agramfort commented on examples/decoding/plot_decoding_time_generalization.py:
```diff
@@ -12,80 +12,81 @@
-raw.filter(1, 45, method='iir')
+# The function implements the method used in:
+# King, Gramfort, Schurger, Naccache & Dehaene, "Two distinct dynamic modes
+# subtend the detection of unexpected sounds", PLOS ONE, 2013
```
this is already in the top docstring.

it's redundant

agramfort replied:
you could add a URL to top docstring

http://www.ncbi.nlm.nih.gov/pubmed/24475052

agramfort commented on examples/decoding/plot_decoding_time_generalization.py:
```diff
@@ -14,78 +14,80 @@
+gat.fit(epochs, y=viz_vs_auditory)
+gat.score(epochs, y=viz_vs_auditory)
+gat.plot_diagonal()  # plot decoding across time (correspond to GAT diagonal)
+gat.plot()  # plot full GAT matrix
```
I get this:

`<GAT | fitted: from -0.053279 s. to 0.399590 s. (0.012939 s. every 0.012939 s.), predicted: 35 trials (predict) on 35 time slice(s) each, scored: True (accuracy_score)>`

some info are redundant also 3 digits after commas is enough. How about :

`<GAT | fit: from -0.053 s. to 0.400 s. (width 0.013 s. step 0.013 s.), predict: 35 trials on 35 time slice(s), scored: True (accuracy_score)>`

agramfort commented on examples/decoding/plot_decoding_time_generalization.py:
```diff
@@ -12,80 +12,81 @@
-event_id = {"faces": 1, "scrambled": 2}
-tmin, tmax = -0.1, 0.5
+# Define events of interest
+events = epochs.events[:, 2]
```
a variable with 1 column should not be called events which is expected to have 3

agramfort commented on mne/decoding/tests/test_time_gen.py:
```diff
@@ -40,3 +44,163 @@ def test_time_generalization():
+    gat = GeneralizationAcrossTime()
+    gat.fit(epochs[0:6])
+    gat.predict(epochs[0:6])
+    assert_raises(ValueError, gat.predict, epochs)
```
I see limited coverage. Only 79%:

mne.decoding.time_gen                338     56    165     48    79%   58, 190, 261, 338, 496, 530, 603, 757, 827-830, 866-889, 898-900, 929-992

a full function in 929-992 is not covered

agramfort commented on mne/decoding/time_gen.py:
```diff
@@ -1,14 +1,997 @@
 from ..io.pick import channel_type, pick_types
 
 
+class DecodingTime(dict):
```
is it meant to be used by average user? if not make it private

agramfort commented on mne/decoding/time_gen.py:
```diff
@@ -1,14 +1,997 @@
+        s = ''
+        if hasattr(self, "estimators_"):
+            # XXX JRK: Shall we print the repr from the DecodingTime() obj
+            # instead?
```
+1

agramfort commented on mne/decoding/time_gen.py:
```diff
@@ -1,14 +1,997 @@
+            test_times_ = DecodingTime()
+        elif isinstance(test_times, dict):
+            # XXX JRK: Should this test whether test_times if a DecodingTime
+            # object or is this sufficient?
```
?

agramfort commented on mne/decoding/time_gen.py:
```diff
@@ -1,14 +1,997 @@
+        Parameters
+        ----------
+        title : str | None
+        Figure title. Defaults to None.
```
indent

agramfort commented on mne/decoding/time_gen.py:
```diff
@@ -1,14 +1,997 @@
+    else:
+        raise RuntimeError('predict_type must be "predict" or "predict_proba" '
+                           'or "decision_function"')
+    y_pred = np.ones((n_trial, n_class, n_clf))
```
we never use the word trial in the rest of mne

agramfort commented on mne/decoding/time_gen.py:
```diff
@@ -1,14 +1,997 @@
+    # Collapse y_pred across folds if necessary (i.e. if independent)
+    if fold > 0:
+        if predict_type == 'predict':
+            y_pred, _ = mode(y_pred, axis=2)
```
please stats.mode as I needed to grep to figure out what you meant by mode

agramfort commented on examples/decoding/plot_decoding_time_generalization.py:
```diff
@@ -12,80 +12,81 @@
-# Set up pick list
-picks = mne.pick_types(raw.info, meg=True, eeg=False, stim=True, eog=True,
-                       ref_meg=False, exclude='bads')
+gat = GeneralizationAcrossTime(predict_mode='cross-validation')
```
add ", n_jobs=1"

to explicit that it can run in parallel

agramfort commented on examples/decoding/plot_decoding_time_generalization.py:
```diff
@@ -12,80 +12,81 @@
-                             shuffle=True, n_jobs=2)
+# As proposed in King & Dehaene (2014) 'Characterizing the dynamics of mental
+# representations: the temporal generalization method', Trends In Cognitive
+# Sciences, 18(4), 203-210.
```
I am tempted to split the example in 2 as I feel this new example much more advanced than the old one and we should fight complexity. That way we would be able to still use SPM data which is nicer to demo the method. Example would then be simple short and nice.

wdyt?

dengemann replied:
+0

dengemann replied:
Let's decide about this later.

agramfort commented:
that's it for me

dengemann commented:
@agramfort is 94% coverage enough for the moment? ...

We're done. All comments addressed.

dengemann commented:
The last commit fixes the sphinx reference. See https://www.dropbox.com/s/94e5x6fv79pnl5y/mne.decoding.GeneralizationAcrossTime.html?dl=0.

Rebase, updated what's new + reference.

agramfort commented on examples/decoding/plot_decoding_time_generalization.py:
```diff
@@ -1,91 +1,87 @@
+viz_vs_auditory_r = (triggers[np.in1d(triggers, (2, 4))] == 4).astype(int)
+
+gat.score(epochs[('AudR', 'VisR')], y=viz_vs_auditory_r)
+gat.plot()
```
you don't like my idea to split the example in 2?

dengemann replied:
Let's leave this to e.g. @kingjr once this is merged.

kingjr replied:
+1  @dengemann

agramfort commented on mne/decoding/time_gen.py:
```diff
@@ -1,14 +1,857 @@
+    # Loop across testing slices
+    y_pred = [list() for _ in range(len(slices))]
+
+    # XXX EHN: This loop should be parallelized in a similar way to fit()
```
?

dengemann replied:
for later to not forget. No XXX will be addressed anymore in this PR.

agramfort commented on mne/viz/tests/test_decoding.py:
```diff
@@ -0,0 +1,68 @@
+
+def test_gat_plot_diagonal():
+    """Test GAT diagonal plot"""
+    gat = _get_data()
```
in this test you call twice get_data which is an overkill of test time I suspect. I'd merge the 2 tests

agramfort commented:
I am attempting to reproduce the old example result with new code

To use ROC-AUC I tried first : scorer='roc_auc' in score method but it failed (works in sklearn btw). Then I read the code and I saw that I need to use predict_type='decision_function' but then it fails with:

ValueError: Data is not binary and pos_label is not specified

@kingjr is the only motivation for predic_type the ability to switch between proba and decision function?

in terms of API would you prefer to use a scorer in init and not use predict_type parameter.

agramfort commented:
I pushed my code if you want to look

kingjr commented:
@agramfort 

1. The error in your example comes from the fact that you used predict_type=decision function, with events values that are different from [0, 1], and this is badly handled by the score function. 

I think this should be corrected at the sklearn level.

I removed the predict_type option so it goes to default (svm + 'probas'). If you want to keep the decision_function parameter, you should redefine the events labels: e.g.

<code>
gat = GeneralizationAcrossTime(predict_mode='cross-validation',
                               predict_type='decision_function')
gat.fit(epochs, y=epochs.events[:,2]-1)
gat.score(epochs, y=epochs.events[:,2]-1)
</code>

2. the predict_type is not only used for scoring. Users often need to retrieve single trial predictions (in which case they may want either discrete predictions (predict), distance from hyperplans (decision_function) or estimated probas (probas). I agree that it may not be ideal in terms of API. However, in my opinion, scikit learn wouldn't use different functions for these, but a predict() for which you can specify what you want kind of prediction you want to make.

dengemann commented:
Travis is happy, who else is?

kingjr commented:
happy too

choldgraf commented:
I'm happy, excited to iterate on this as a part of mne-python!

agramfort commented:
@kingjr ok but `gat.score(epochs)` should work as much as possible out of the box.

the old example used ROC-AUC and I wanted to keep it like this. Experience shows that it was not immediate.

I don't want to be a pain but I am just fighting to maintain MNE as obvious to use as possible. Or at least as "obvious" as it can be...

agramfort commented:
if you're all happy I give up. We can iterate in master I guess... The new example will be different though

kingjr commented:
@agramfort I understand. 

How about we deactivate the decision_function for now so that we can deal with it later?

dengemann commented:
Let's adress this later, sklearn is beasty with its ROC-AUC api

> On 10 Mar 2015, at 22:02, Alexandre Gramfort <notifications@github.com> wrote:
> 
> @kingjr ok but gat.score(epochs) should work as much as possible out of the box.
> 
> the old example used ROC-AUC and I wanted to keep it like this. Experience shows that it was not immediate.
> 
> I don't want to be a pain but I am just fighting to maintain MNE as obvious to use as possible. Or at least as "obvious" as it can be...
> 
> —
> Reply to this email directly or view it on GitHub.
>

dengemann commented:
Could we please minimize / avoid api changes at this stage. For each of them we tend to spend days, as experience shows. 

> On 10 Mar 2015, at 22:06, J-R King <notifications@github.com> wrote:
> 
> @agramfort I understand.
> 
> How about we deactivate the decision_function for now so that we can deal with it later?
> 
> —
> Reply to this email directly or view it on GitHub.
>

choldgraf commented:
Actually I do think @agramfort makes a good point about consistency and usability. I was thinking iterate as in improve and add extra functionality, but it seems unwise to push something to master where there's already a plan to change its core functionality (e.g., changing the way that you define a scorer or whatever). If I were a new user of MNE-python, and was using it because I wanted to try out this time-gen method, it might be frustrating to encounter strangeness because of this kind of thing.

choldgraf commented:
Maybe with the proper documentation and/or examples this wouldn't be so much of a problem. You guys mentioned that the way you handle examples on the website is about to change as well, no?

dengemann commented:
This is why it's the dev branch, it is expected to slightly change. This one I would not see as change but as future enhancement, i.e., handle sklearn roc auc label API internally.

> On 10 Mar 2015, at 22:09, Chris Holdgraf <notifications@github.com> wrote:
> 
> Maybe with the proper documentation and/or examples this wouldn't be so much of a problem. You guys mentioned that the way you handle examples on the website is about to change as well, no?
> 
> —
> Reply to this email directly or view it on GitHub.
>

agramfort commented:
as we want to release soon it would be a subsequent PR really imminently

choldgraf commented:
That is a good point, I maintain my +1 :)

agramfort commented:
you don't have commit rights but I accept your +1 :)

I guess we just need another one...

Eric89GXL commented:
I am a +0. It makes me worry that @agramfort couldn't get it to work in some easy way, but this has also been open a painfully long time.

dengemann commented:
I take that point. We should handle the sklearn pecularities with AUC internally, which we need to do either way if sklearn improves on that in the future, for backwards compatibility reasons. If you agree we will merge it once this is addressed. Ok?

> On 10 Mar 2015, at 22:27, Eric Larson <notifications@github.com> wrote:
> 
> I am a +0. It makes me worry that @agramfort couldn't get it to work in some easy way, but this has also been open a painfully long time.
> 
> —
> Reply to this email directly or view it on GitHub.
>

Eric89GXL commented:
fine by me

dengemann commented:
@Eric89GXL @agramfort @kingjr @choldgraf @cmoutard

The API probelm that we have looks as follows to me. The sklearn auc socrer requires a definition of the target class or binary labels, {0, 1}. We can handle it internally, but it will be implicit. This may be convenient but also a potential source of confusion. Ultimately users need to have an idea about what tool they are using. The options I see:

1) use accuracy by default as score
2) keep as is and add a warning / error
3) assume that higher values in the 3rd column of the events array are ones, lower zeros, for data with 2 labels. Likewise, assume that e.g. the first epoch in in a list of epochs represetns the target class.

Please be verbose.

kingjr commented:
@dengemann 
Regarding point 1. 
- The default is currently clf=SVC, predict_type='predict' scorer='accuracy'. 
- If the user manually changes predict_type='decision_function', then the scorer='auc'. 
- I don't think it would be appropriate to force an accuracy for predict_type='decision_function', as it would then be identical to predict_type='predict', and would require an extra processing step (transforming the distance from the hyperplan into a categorical value).

I agree with 2) and 3).

dengemann commented:
I would prefer 2) as 3) makes too strong assumptions and 1) is a bit silly, as you often want an AUC if using probability outputs.

agramfort commented:
to deal with y that is not 0, 1 sklearn uses a LabelEncoder see

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/logistic.py#L547

it would convert targets [2, 3] to [0, 1]

I think It would fix the problem.

dengemann commented:
Yes, but what I don't like about it is, that event ids are arbitrary and this makes e.g. right the target, not left for random reasons. If we go that way, we need verbosity.

agramfort commented:
+1 for verbose. it's ordered by increasing event id values

kingjr commented:
ok +1, i'll check this out now

kingjr commented:
The main issue I see with LabelEncoder is that this only applies to categorical learning, and thus excludes SVR, LogisticReg etc. 

It's tricky to detect whether the data is categorical (you can do a regression on integers values so we can't use that; we can't use the number of different values in y either, as it can widely vary across studies).

Any thoughts?

agramfort commented:
use sklearn.base.is_classifier to test if you have a classifier or a regressor

dengemann commented:
> use sklearn.base.is_classifier to test if you have a classifier or a regressor

That sounds like the way to go.

kingjr commented:
@agramfort the code now works with your previous example (predict_type='decision_function'), but I think I still removed this parameter from the example for sake of simplicity.

agramfort commented on examples/decoding/plot_decoding_time_generalization.py:
```diff
@@ -1,91 +1,53 @@
-plt.axhline(0, color='k')
-plt.colorbar()
+# Define decoder. The decision_function is employed to use AUC for scoring
+gat = GeneralizationAcrossTime(predict_mode='cross-validation', n_jobs=1)
```
doc line above is not up to date. I'd rather advertise ROC AUC in example as to be it's more powerful to see effects

kingjr commented:
on my computer
```%time test_generalization_across_time()
CPU times: user 12.4 s, sys: 168 ms, total: 12.6 s
Wall time: 8.29 s```

ok for the example.

dengemann commented:
Her on my machine.

```
CPU times: user 6.36 s, sys: 33 ms, total: 6.4 s
Wall time: 5.24 s
```

dengemann commented:
cut it down by 2 seconds. Good to go now from my side.

agramfort commented:
@dengemann merge when you're happy

agramfort commented:
nice team work everyone !

dengemann commented:
Thanks everyone for your contributions!!

> On 11 Mar 2015, at 17:44, Alexandre Gramfort <notifications@github.com> wrote:
> 
> @dengemann merge when you're happy
> 
> —
> Reply to this email directly or view it on GitHub.
>

kingjr commented:
Great, thank you very much indeed!
